### PR TITLE
libllvm: add flang

### PR DESCRIPTION
**Pull request by Grimler91**
Continuation of https://github.com/termux/termux-packages/pull/6568

### CODE REVIEW

**Comment by thunder-coding** [comment]
Seems like the CI isn't powerful enough to compile flang. Even with `TERMUX_MAKE_PROCESSES=1` it fails.

**Comment by Grimler91** [comment]
Locally it still fails with 
```
/bin/sh: 1: /home/builder/.termux-build/libllvm/build/NATIVE/bin/mlir-tblgen: Exec format error
```
when building things in `/home/builder/.termux-build/libllvm/src/flang/lib/`

**Comment by buttaface** [comment]
Oh yeah, that was the same problem the original pull had. Since you are now building and passing `mlir-tblgen` in separately, can you just disable that entire `NATIVE` build step? It isn't used for anything.

**Comment by Grimler91** [comment]
> Oh yeah, that was the same problem the original pull had. Since you are now building and passing `mlir-tblgen` in separately, can you just disable that entire `NATIVE` build step? It isn't used for anything.

Yes, right, we discussed it already :) That could work, just have to find where to disable it..

**Comment by buttaface** [comment]
I think it might be [built here](https://github.com/llvm/llvm-project/blob/main/llvm/cmake/modules/TableGen.cmake#L154): can you try changing that to `if(NOT LLVM_USE_HOST_TOOLS)`?

**Comment by Grimler91** [comment]
@buttaface thanks, building now to see if it works

**Comment by Grimler91** [comment]
flang build now instead tries to use cross-compiled llvm-tblgen:

```
[99/319] Building Attributes.inc...                                                                                                                                                                                                           
FAILED: include/llvm/IR/Attributes.inc /home/builder/.termux-build/libllvm/build/NATIVE/include/llvm/IR/Attributes.inc                                                                                                                        
cd /home/builder/.termux-build/libllvm/build/NATIVE && /home/builder/.termux-build/libllvm/build/NATIVE/bin/llvm-tblgen -gen-attrs -I /home/builder/.termux-build/libllvm/src/llvm/include/llvm/IR -I/home/builder/.termux-build/libllvm/build/NATIVE/include -I/home/builder/.termux-build/libllvm/src/llvm/include /home/builder/.termux-build/libllvm/src/llvm/include/llvm/IR/Attributes.td --write-if-changed -o include/llvm/IR/Attributes.inc -d include/llvm/IR/Attributes.inc.d    
/bin/sh: 1: /home/builder/.termux-build/libllvm/build/NATIVE/bin/llvm-tblgen: Exec format error                                                                                                                                               
[108/319] Building CXX object tools/mlir/tools/mlir-tblgen/CMakeFiles/mlir-tblgen.dir/OpDefinitionsGen.cpp.o                                                                                                                                  
ninja: build stopped: subcommand failed. 
```

In the NATIVE/bin directory we have:

```
$ file ~/.termux-build/libllvm/build/NATIVE/bin/*
/home/builder/.termux-build/libllvm/build/NATIVE/bin/flang:       Bourne-Again shell script, ASCII text executable
/home/builder/.termux-build/libllvm/build/NATIVE/bin/lldb-dotest: Python script, ASCII text executable
/home/builder/.termux-build/libllvm/build/NATIVE/bin/lldb-repro:  Python script, ASCII text executable
/home/builder/.termux-build/libllvm/build/NATIVE/bin/llvm-config: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /system/bin/linker64, not stripped
/home/builder/.termux-build/libllvm/build/NATIVE/bin/llvm-lit:    Python script, ASCII text executable
/home/builder/.termux-build/libllvm/build/NATIVE/bin/llvm-tblgen: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /system/bin/linker64, not stripped
```

**Comment by buttaface** [comment]
> flang build now instead tries to use cross-compiled llvm-tblgen

I think that's just a standard LLVM file it's generating.

Looking into it more, I think the original way you had it was the right approach.The problem is that while [LLVM, clang, and lldb support passing in host tablegen executables](https://github.com/llvm/llvm-project/blob/main/llvm/docs/HowToCrossCompileLLVM.rst#configuring-cmake), MLIR tablegen probably doesn't support that yet and defaults to trying to use that broken `NATIVE/` directory instead.

I suggest you revert this last commit- don't worry about disabling the `NATIVE/` build unless it's not building- and look into patching the CMake config to use our host mlir-tblgen properly.

**Comment by Grimler91** [comment]
> I suggest you revert this last commit- don't worry about disabling the NATIVE/ build unless it's not building- and look into patching the CMake config to use our host mlir-tblgen properly.

Soudns like a plan. With latest fixup commit we no longer generate NATIVE/bin/llvm-tblgen, but instead it tries to run `/home/builder/.termux-build/libllvm/build/bin/mlir-tblgen`:
```
[4217/5381] cd /home/builder/.termux-build/libllvm/build && /home/builder/.termux-build/libllvm/build/bin/mlir-tblgen -gen-op-decls -I /home/builder/.termux-build/libllvm/src/flang/include/flang/Optimizer/Dialect -I/home/builder/.termux-build/libllvm/src/flang/include -I/home/builder/.termux-build/libllvm/build/tools/flang/include -I/home/builder/.termux-build/libllvm/build/include -I/home/builder/.termux-build/libllvm/src/llvm/include -I/home/builder/.termux-build/libllvm/src/llvm/../mlir/include -I/home/builder/.termux-build/libllvm/build/tools/mlir/include -I/home/builder/.termux-build/libllvm/build/tools/clang/include -I/home/builder/.termux-build/libllvm/src/llvm/../clang/include /home/builder/.termux-build/libllvm/src/flang/include/flang/Optimizer/Dialect/FIROps.td --write-if-changed -o tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc -d tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc.d
FAILED: tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc /home/builder/.termux-build/libllvm/build/tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc 
cd /home/builder/.termux-build/libllvm/build && /home/builder/.termux-build/libllvm/build/bin/mlir-tblgen -gen-op-decls -I /home/builder/.termux-build/libllvm/src/flang/include/flang/Optimizer/Dialect -I/home/builder/.termux-build/libllvm/src/flang/include -I/home/builder/.termux-build/libllvm/build/tools/flang/include -I/home/builder/.termux-build/libllvm/build/include -I/home/builder/.termux-build/libllvm/src/llvm/include -I/home/builder/.termux-build/libllvm/src/llvm/../mlir/include -I/home/builder/.termux-build/libllvm/build/tools/mlir/include -I/home/builder/.termux-build/libllvm/build/tools/clang/include -I/home/builder/.termux-build/libllvm/src/llvm/../clang/include /home/builder/.termux-build/libllvm/src/flang/include/flang/Optimizer/Dialect/FIROps.td --write-if-changed -o tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc -d tools/flang/include/flang/Optimizer/Dialect/FIROps.h.inc.d
/bin/sh: 1: /home/builder/.termux-build/libllvm/build/bin/mlir-tblgen: Exec format error
```

Might be a better strategy to look into actually making flang support cross-compilation in a similar way as clang and friends

**Comment by buttaface** [comment]
> NATIVE/bin/llvm-tblgen, but instead it tries to run /home/builder/.termux-build/libllvm/build/bin/mlir-tblgen

OK, you just need to swap that out for the host-build mlir-tblgen and it should work.

**Comment by Grimler91** [comment]
Reviving the patch to sed path to mlir-tblgen lets us proceed further indeed. Now we get a more normal build error:

```
[5506/7789] /home/builder/.termux-build/_cache/android-r23b-api-24-v1/bin/clang++ --target=armv7-none-linux-androideabi --gcc-toolchain=/home/builder/.termux-build/_cache/android-r23b-api-24-v1 --sysroot=/home/builder/.termux-build/_cache/android-r23b-api-24-v1/sysroot -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/builder/.termux-build/libllvm/build/tools/flang/lib/Evaluate -I/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate -I/home/builder/.termux-build/libllvm/src/flang/include -I/home/builder/.termux-build/libllvm/build/tools/flang/include -I/home/builder/.termux-build/libllvm/build/include -I/home/builder/.termux-build/libllvm/src/llvm/include -isystem /home/builder/.termux-build/libllvm/src/llvm/../mlir/include -isystem /home/builder/.termux-build/libllvm/build/tools/mlir/include -isystem /home/builder/.termux-build/libllvm/build/tools/clang/include -isystem /home/builder/.termux-build/libllvm/src/llvm/../clang/include -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -Oz --target=armv7a-linux-androideabi24  -I/data/data/com.termux/files/usr/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-deprecated-copy -Wno-string-conversion -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -O3 -DNDEBUG -fPIC  -fno-exceptions -fno-rtti -std=c++17 -MD -MT tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o -MF tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o.d -o tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o -c /home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp
FAILED: tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o 
/home/builder/.termux-build/_cache/android-r23b-api-24-v1/bin/clang++ --target=armv7-none-linux-androideabi --gcc-toolchain=/home/builder/.termux-build/_cache/android-r23b-api-24-v1 --sysroot=/home/builder/.termux-build/_cache/android-r23b-api-24-v1/sysroot -DGTEST_HAS_RTTI=0 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/builder/.termux-build/libllvm/build/tools/flang/lib/Evaluate -I/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate -I/home/builder/.termux-build/libllvm/src/flang/include -I/home/builder/.termux-build/libllvm/build/tools/flang/include -I/home/builder/.termux-build/libllvm/build/include -I/home/builder/.termux-build/libllvm/src/llvm/include -isystem /home/builder/.termux-build/libllvm/src/llvm/../mlir/include -isystem /home/builder/.termux-build/libllvm/build/tools/mlir/include -isystem /home/builder/.termux-build/libllvm/build/tools/clang/include -isystem /home/builder/.termux-build/libllvm/src/llvm/../clang/include -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -Oz --target=armv7a-linux-androideabi24  -I/data/data/com.termux/files/usr/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-deprecated-copy -Wno-string-conversion -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -O3 -DNDEBUG -fPIC  -fno-exceptions -fno-rtti -std=c++17 -MD -MT tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o -MF tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o.d -o tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o -c /home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: error: non-constant-expression cannot be narrowed from type 'long long' to 'std::size_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:322:1: note: in instantiation of member function 'Fortran::evaluate::Constant<Fortran::evaluate::Type<TypeCategory::Character, 1>>::CopyFrom' requested here
INSTANTIATE_CONSTANT_TEMPLATES
^
/home/builder/.termux-build/libllvm/src/flang/include/flang/Evaluate/constant.h:236:42: note: expanded from macro 'INSTANTIATE_CONSTANT_TEMPLATES'
  FOR_EACH_INTRINSIC_KIND(template class Constant, )
                                         ^
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: note: insert an explicit cast to silence this issue
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                             static_cast<size_t>(                  )
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: error: non-constant-expression cannot be narrowed from type 'long long' to 'std::size_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:322:1: note: in instantiation of member function 'Fortran::evaluate::Constant<Fortran::evaluate::Type<TypeCategory::Character, 2>>::CopyFrom' requested here
INSTANTIATE_CONSTANT_TEMPLATES
^
/home/builder/.termux-build/libllvm/src/flang/include/flang/Evaluate/constant.h:236:42: note: expanded from macro 'INSTANTIATE_CONSTANT_TEMPLATES'
  FOR_EACH_INTRINSIC_KIND(template class Constant, )
                                         ^
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: note: insert an explicit cast to silence this issue
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                             static_cast<size_t>(                  )
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: error: non-constant-expression cannot be narrowed from type 'long long' to 'std::size_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:322:1: note: in instantiation of member function 'Fortran::evaluate::Constant<Fortran::evaluate::Type<TypeCategory::Character, 4>>::CopyFrom' requested here
INSTANTIATE_CONSTANT_TEMPLATES
^
/home/builder/.termux-build/libllvm/src/flang/include/flang/Evaluate/constant.h:236:42: note: expanded from macro 'INSTANTIATE_CONSTANT_TEMPLATES'
  FOR_EACH_INTRINSIC_KIND(template class Constant, )
                                         ^
/home/builder/.termux-build/libllvm/src/flang/lib/Evaluate/constant.cpp:254:30: note: insert an explicit cast to silence this issue
    std::size_t elementBytes{length_ * sizeof(decltype(values_[0]))};
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                             static_cast<size_t>(                  )
3 errors generated.
```
which shouldn't be too hard to solve

**Comment by buttaface** [comment]
Heh, the CI ran out of space on the aarch64 host build because it built the erlang package first, need to rebase this to avoid that.

File: packages/libllvm/build.sh [hunk 1]
```diff
@@ -57,13 +58,15 @@ TERMUX_PKG_HAS_DEBUG=false
...
 
 termux_step_host_build() {
 	termux_setup_cmake
 	termux_setup_ninja
 
-	cmake -G Ninja -DLLVM_ENABLE_PROJECTS='clang;lldb' $TERMUX_PKG_SRCDIR/llvm
-	ninja -j $TERMUX_MAKE_PROCESSES clang-tblgen lldb-tblgen llvm-tblgen
+	cmake -G Ninja -DLLVM_ENABLE_PROJECTS='clang;flang;lldb;mlir' $TERMUX_PKG_SRCDIR/llvm
+	ninja -j $TERMUX_MAKE_PROCESSES clang-tblgen lldb-tblgen llvm-tblgen \
+				        mlir-tblgen mlir-linalg-ods-gen f18
```
Comment (buttaface) [inline packages/libllvm/build.sh:69]: Do you need to build `f18` here? I don't see this host f18 passed to the cross-compile.

File: packages/libllvm/build.sh [hunk 2]
```diff
@@ -93,6 +96,8 @@ termux_step_pre_configure() {
 
 termux_step_post_configure() {
 	TERMUX_PKG_SRCDIR=$TERMUX_SRCDIR_SAVE
+	# add hostbuild dir to path so mlir-tblgen is found
+	export PATH=$TERMUX_PKG_HOSTBUILD_DIR/bin:$PATH
```
Comment (buttaface) [inline packages/libllvm/build.sh:100]: No longer needed if you're patching the path in below?

File: packages/libllvm/build.sh (hunk 1 above)
Comment (Grimler91) [reply packages/libllvm/build.sh:69]: Will try without

File: packages/libllvm/build.sh (hunk 2 above)
Comment (Grimler91) [reply packages/libllvm/build.sh:100]: It seem to be needed, for whatever reason, även with the patch it runs just `mlir-tblgen` from PATH

**Comment by buttaface** [comment]
Timed out on CI, but at least no build errors. Re-enabled using both CI cores and rebased, let's see if it finishes now.

**Comment by buttaface** [comment]
Looks like more typecasting issues on 32-bit platforms.

**Comment by Grimler91** [comment]
As well as it still trying to run llvm-tblgen from NATIVE/bin dir: https://github.com/termux/termux-packages/runs/4089337018?check_suite_focus=true#step:3:10816

**Comment by buttaface** [comment]
The issue is that it's trying to [build a "native mlir-linalg-ods-gen"](https://github.com/termux/termux-packages/runs/4089337018?check_suite_focus=true#step:3:10729): you'll need to disable that too, since you already built it in the hostbuild step.

**Comment by buttaface** [comment]
Btw, I have no interest in this flang compiler, which is why I'm not digging deeper. I'm happy to advise anyone who wants to get this built, but I will not be working on this port.

**Comment by Grimler91** [comment]
@buttaface sure, I will continue look at it, and test it, when I have some time. Thanks for your help so far!

**Comment by Grimler91** [comment]
Builds for at least for aarch64 now.
flang needs a dependency on uuidgen as well. For uuidgen I guess we need to [enable libuuid](https://github.com/termux/termux-packages/blob/master/packages/util-linux/build.sh#L23) for util-linux, and remove the [libuuid package](https://github.com/termux/termux-packages/tree/master/packages/libuuid). Would require rebuilding all packages that currently depend on libuuid as well.

**Comment by zorro** [comment]
@Grimler91 have a look at https://lfortran.org/ . Building without any backends seems easy. But enabling ```LLVM``` backend requires some work

**Comment by Grimler91** [comment]
> Building without any backends seems easy. 

You and I seem to have different definitions of easy, considering all the problems in this (and the previous) PR :)

> But enabling `LLVM` backend requires some work

Yes, flang as enabled here is only the frontend (which I wasn't aware of when I started looking at it). There are other compilers (not only lfortran) that could be used as backend as well: https://fortran-lang.org/compilers/. After we merge this we can compile fortran using gfortran from [@its-pointless repo](https://github.com/its-pointless/its-pointless.github.io/), until llvm comes with its own compiler (or we add another one).

**Comment by jalomic** [comment]
Any news ?

**Comment by Grimler91** [comment]
> No space left on device

Aaaargh, it can't even finish the host build step..

**Comment by thunder-coding** [comment]
> > No space left on device
> 
> Aaaargh, it can't even finish the host build step..

I have configured the CI to now free disk space when building libllvm, hope it helps. (Also rebased against master as usual)

**Pull request closed without merging**